신규 유저 브로드캐스트 시 성능 저하(24s), OOM 위험 및 낙관적 락 충돌 이슈

## 🐞 버그 내용
- 어떤 문제가 발생했는지 구체적으로 작성해주세요.

2025-11-29T18:32:44.391+09:00  WARN 1 --- [     dispatch-5] [                                                 ] c.d.n.service.NotificationEventListener  : [성능 경고] 848명의 모든 사용자를 메모리에 로드했습니다. 사용자 수가 많을 경우 OutOfMemoryError가 발생할 수 있습니다.
2025-11-29T18:32:44.398+09:00  WARN 1 --- [     dispatch-4] [                                                 ] c.d.n.service.NotificationEventListener  : [성능 경고] 848명의 모든 사용자를 메모리에 로드했습니다. 사용자 수가 많을 경우 OutOfMemoryError가 발생할 수 있습니다.
2025-11-29T18:32:54.603+09:00  WARN 1 --- [nio-8080-exec-8] [692abdc6d352d0e786959341d658e7e5-8c6e9be6b40e412a] core.global.controller.UserController    : 사용자의 요청한 리프레쉬 토큰 : eyJhbGciOiJIUzUxMiJ9.eyJzdWIiOiIxMjEwIiwianRpIjoiNTJlZDEwOTMtYjIxMS00M2Q4LWE2MzMtOTJlMmVkNGQ5ZTc3IiwiaWF0IjoxNzY0NDA4NzI2LCJleHAiOjE3NjUwMTM1MjZ9.A-1Hb3DHQroXs-Td7uvAtvbYH-IvhtnNNr5kUfue2DOvssVsiwdF1YEESC-mYeWQHMsoUQfxn8leBCJaz2H6oA
2025-11-29T18:32:54.603+09:00  WARN 1 --- [nio-8080-exec-8] [692abdc6d352d0e786959341d658e7e5-8c6e9be6b40e412a] core.global.controller.UserController    : 레디스의 요청한 리프레쉬 토큰 : eyJhbGciOiJIUzUxMiJ9.eyJzdWIiOiIxMjEwIiwianRpIjoiNTJlZDEwOTMtYjIxMS00M2Q4LWE2MzMtOTJlMmVkNGQ5ZTc3IiwiaWF0IjoxNzY0NDA4NzI2LCJleHAiOjE3NjUwMTM1MjZ9.A-1Hb3DHQroXs-Td7uvAtvbYH-IvhtnNNr5kUfue2DOvssVsiwdF1YEESC-mYeWQHMsoUQfxn8leBCJaz2H6oA
2025-11-29T18:33:02.346+09:00 ERROR 1 --- [     dispatch-5] [                                                 ] c.d.n.service.NotificationEventListener  : 사용자 ID 1006에게 신규 유저 알림 발송 중 오류 발생
org.springframework.orm.ObjectOptimisticLockingFailureException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect): [core.domain.userdevicetoken.entity.UserDeviceToken#476]
        at org.springframework.orm.jpa.vendor.HibernateJpaDialect.convertHibernateAccessException(HibernateJpaDialect.java:329) ~[spring-orm-6.2.9.jar!/:6.2.9]
        at org.springframework.orm.jpa.vendor.HibernateJpaDialect.convertHibernateAccessException(HibernateJpaDialect.java:256) ~[spring-orm-6.2.9.jar!/:6.2.9]
        at org.springframework.orm.jpa.vendor.HibernateJpaDialect.translateExceptionIfPossible(HibernateJpaDialect.java:244) ~[spring-orm-6.2.9.jar!/:6.2.9]
        at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.translateExceptionIfPossible(AbstractEntityManagerFactoryBean.java:560) ~[spring-orm-6.2.9.jar!/:6.2.9]
        at org.springframework.dao.support.ChainedPersistenceExceptionTranslator.translateExceptionIfPossible(ChainedPersistenceExceptionTranslator.java:61) ~[spring-tx-6.2.9.jar!/:6.2.9]
        at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:160) ~[spring-tx-6.2.9.jar!/:6.2.9]
Caused by: org.hibernate.StaleObjectStateException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect): [core.domain.userdevicetoken.entity.UserDeviceToken#476]
        at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:138) ~[spring-tx-6.2.9.jar!/:6.2.9]
2025-11-29T18:33:02.357+09:00 ERROR 1 --- [     dispatch-5] [                                                 ] c.d.n.service.NotificationEventListener  : 사용자 ID 990에게 신규 유저 알림 발송 중 오류 발생
org.springframework.orm.ObjectOptimisticLockingFailureException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect): [core.domain.userdevicetoken.entity.UserDeviceToken#476]
        at org.springframework.orm.jpa.vendor.HibernateJpaDialect.convertHibernateAccessException(HibernateJpaDialect.java:329) ~[spring-orm-6.2.9.jar!/:6.2.9]
        at org.springframework.orm.jpa.vendor.HibernateJpaDialect.convertHibernateAccessException(HibernateJpaDialect.java:256) ~[spring-orm-6.2.9.jar!/:6.2.9]
        at org.springframework.orm.jpa.vendor.HibernateJpaDialect.translateExceptionIfPossible(HibernateJpaDialect.java:244) ~[spring-orm-6.2.9.jar!/:6.2.9]
        at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.translateExceptionIfPossible(AbstractEntityManagerFactoryBean.java:560) ~[spring-orm-6.2.9.jar!/:6.2.9]
        at org.springframework.dao.support.ChainedPersistenceExceptionTranslator.translateExceptionIfPossible(ChainedPersistenceExceptionTranslator.java:61) ~[spring-tx-6.2.9.jar!/:6.2.9]
        at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:160) ~[spring-tx-6.2.9.jar!/:6.2.9]
Caused by: org.hibernate.StaleObjectStateException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect): [core.domain.userdevicetoken.entity.UserDeviceToken#476]
        at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:138) ~[spring-tx-6.2.9.jar!/:6.2.9]
2025-11-29T18:33:02.368+09:00 ERROR 1 --- [     dispatch-5] [                                                 ] c.d.n.service.NotificationEventListener  : 사용자 ID 1081에게 신규 유저 알림 발송 중 오류 발생
org.springframework.orm.ObjectOptimisticLockingFailureException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect): [core.domain.userdevicetoken.entity.UserDeviceToken#476]
        at org.springframework.orm.jpa.vendor.HibernateJpaDialect.convertHibernateAccessException(HibernateJpaDialect.java:329) ~[spring-orm-6.2.9.jar!/:6.2.9]
        at org.springframework.orm.jpa.vendor.HibernateJpaDialect.convertHibernateAccessException(HibernateJpaDialect.java:256) ~[spring-orm-6.2.9.jar!/:6.2.9]
        at org.springframework.orm.jpa.vendor.HibernateJpaDialect.translateExceptionIfPossible(HibernateJpaDialect.java:244) ~[spring-orm-6.2.9.jar!/:6.2.9]
        at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.translateExceptionIfPossible(AbstractEntityManagerFactoryBean.java:560) ~[spring-orm-6.2.9.jar!/:6.2.9]
        at org.springframework.dao.support.ChainedPersistenceExceptionTranslator.translateExceptionIfPossible(ChainedPersistenceExceptionTranslator.java:61) ~[spring-tx-6.2.9.jar!/:6.2.9]
        at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:160) ~[spring-tx-6.2.9.jar!/:6.2.9]
Caused by: org.hibernate.StaleObjectStateException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect): [core.domain.userdevicetoken.entity.UserDeviceToken#476]
        at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:138) ~[spring-tx-6.2.9.jar!/:6.2.9]
2025-11-29T18:33:02.376+09:00 ERROR 1 --- [     dispatch-5] [                                                 ] c.d.n.service.NotificationEventListener  : 사용자 ID 991에게 신규 유저 알림 발송 중 오류 발생
org.springframework.orm.ObjectOptimisticLockingFailureException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect): [core.domain.userdevicetoken.entity.UserDeviceToken#476]
        at org.springframework.orm.jpa.vendor.HibernateJpaDialect.convertHibernateAccessException(HibernateJpaDialect.java:329) ~[spring-orm-6.2.9.jar!/:6.2.9]
        at org.springframework.orm.jpa.vendor.HibernateJpaDialect.convertHibernateAccessException(HibernateJpaDialect.java:256) ~[spring-orm-6.2.9.jar!/:6.2.9]
        at org.springframework.orm.jpa.vendor.HibernateJpaDialect.translateExceptionIfPossible(HibernateJpaDialect.java:244) ~[spring-orm-6.2.9.jar!/:6.2.9]
        at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.translateExceptionIfPossible(AbstractEntityManagerFactoryBean.java:560) ~[spring-orm-6.2.9.jar!/:6.2.9]
        at org.springframework.dao.support.ChainedPersistenceExceptionTranslator.translateExceptionIfPossible(ChainedPersistenceExceptionTranslator.java:61) ~[spring-tx-6.2.9.jar!/:6.2.9]
        at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:160) ~[spring-tx-6.2.9.jar!/:6.2.9]
Caused by: org.hibernate.StaleObjectStateException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect): [core.domain.userdevicetoken.entity.UserDeviceToken#476]
        at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:138) ~[spring-tx-6.2.9.jar!/:6.2.9]
2025-11-29T18:33:02.383+09:00 ERROR 1 --- [     dispatch-5] [                                                 ] c.d.n.service.NotificationEventListener  : 사용자 ID 975에게 신규 유저 알림 발송 중 오류 발생
org.springframework.orm.ObjectOptimisticLockingFailureException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect): [core.domain.userdevicetoken.entity.UserDeviceToken#476]
        at org.springframework.orm.jpa.vendor.HibernateJpaDialect.convertHibernateAccessException(HibernateJpaDialect.java:329) ~[spring-orm-6.2.9.jar!/:6.2.9]
        at org.springframework.orm.jpa.vendor.HibernateJpaDialect.convertHibernateAccessException(HibernateJpaDialect.java:256) ~[spring-orm-6.2.9.jar!/:6.2.9]
        at org.springframework.orm.jpa.vendor.HibernateJpaDialect.translateExceptionIfPossible(HibernateJpaDialect.java:244) ~[spring-orm-6.2.9.jar!/:6.2.9]
        at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.translateExceptionIfPossible(AbstractEntityManagerFactoryBean.java:560) ~[spring-orm-6.2.9.jar!/:6.2.9]
        at org.springframework.dao.support.ChainedPersistenceExceptionTranslator.translateExceptionIfPossible(ChainedPersistenceExceptionTranslator.java:61) ~[spring-tx-6.2.9.jar!/:6.2.9]
        at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:160) ~[spring-tx-6.2.9.jar!/:6.2.9]
Caused by: org.hibernate.StaleObjectStateException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect): [core.domain.userdevicetoken.entity.UserDeviceToken#476]
        at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:138) ~[spring-tx-6.2.9.jar!/:6.2.9]
2025-11-29T18:33:02.400+09:00 ERROR 1 --- [     dispatch-5] [                                                 ] c.d.n.service.NotificationEventListener  : 사용자 ID 992에게 신규 유저 알림 발송 중 오류 발생

---

## 🧩 예상 원인
- 문제의 원인이 무엇일지 추정합니다.
- (예: 검색 시 상태 초기화 로직이 잘못 작동하여 북마크 상태가 초기화됨)

---

## ✅ 재현 경로
1. 페이지 접속
2. 버튼 클릭 등 버그를 재현할 수 있는 순서를 단계별로 작성
3. 발생한 문제와 실제 결과, 기대한 결과를 명시

---

## 📌 이슈 내용
- 어떤 기능/버그에 대한 작업인지 간략히 요약합니다.

---

## ✅ 작업 내용
- [ ] 작업 1
- [ ] 작업 2

---

## 🔗 관련 이슈/PR
- (선택사항)

--- a/src/main/java/core/domain/chat/dto/MessageSentEvent.java
+++ b/src/main/java/core/domain/chat/dto/MessageSentEvent.java
@@ -9,5 +9,6 @@ import java.util.List;
 public record MessageSentEvent(
         ChatMessageResponse messageResponse,    // 전송된 메시지 내용
         List<Long> recipientIds,                // 수신자 ID 목록
-        ChatRoomSummaryResponse roomSummary     // 갱신될 채팅방 요약 정보
+        ChatRoomSummaryResponse roomSummary   ,  // 갱신될 채팅방 요약 정보
+        long startTime //
 ) {}

--- a/src/main/java/core/domain/chat/service/ChatEventListener.java
+++ b/src/main/java/core/domain/chat/service/ChatEventListener.java
@@ -100,7 +100,12 @@ public class ChatEventListener {
                 eventPublisher.publishEvent(notificationEvent);
             }
         });
+        long currentTime = System.currentTimeMillis();
+        long totalDuration = currentTime - event.startTime();
 
+        log.info("🚀 [E2E Performance] Message {} broadcast complete.", message.id());
+        log.info("   - Recipients: {} users", event.recipientIds().size());
+        log.info("   - Total E2E Latency: {} ms (Service + DB Commit + Async Wait + Socket Push)", totalDuration);
         log.info("Message {} broadcasted via Parallel Stream to {} recipients (DB Query Skipped)",
                 message.id(), event.recipientIds().size());
     }

--- a/src/main/java/core/domain/chat/service/ChatMessageService.java
+++ b/src/main/java/core/domain/chat/service/ChatMessageService.java
@@ -89,7 +89,7 @@ public class ChatMessageService {
         long startTime = System.currentTimeMillis();
 
         try {
-
+            long absoluteStartTime = System.currentTimeMillis();
             // 1. 메시지 저장
             ChatMessage savedMessage = this.saveMessage(req.roomId(), req.senderId(), req.content());
             ChatRoom chatRoom = savedMessage.getChatRoom();
@@ -171,18 +171,16 @@ public class ChatMessageService {
                 );
 
                 ChatRoomSummaryResponse summary = buildChatRoomSummaryResponse(chatRoom.getId(), sender.getId());
-                eventPublisher.publishEvent(new MessageSentEvent(messageResponse, recipientIds, summary));
+                eventPublisher.publishEvent(new MessageSentEvent(messageResponse, recipientIds, summary,absoluteStartTime));
 
-                chatMetrics.onMessageSent("text", true);
-                chatMetrics.recordDelivery(startTime);
+                //chatMetrics.onMessageSent("text", true);
+                //chatMetrics.recordDelivery(startTime);
 
                 long endTime = System.currentTimeMillis();
-                log.debug("Processed message for room {} in {}ms (Recipients: {})",
-                        req.roomId(), (endTime - startTime), chatRoom.getParticipants().size());
             }
         } catch (Exception e) {
-            chatMetrics.onMessageSent("text", false);
-            chatMetrics.recordDelivery(startTime);
+           // chatMetrics.onMessageSent("text", false);
+          //  chatMetrics.recordDelivery(startTime);
 
             throw e;
         }
@@ -350,6 +348,7 @@ public class ChatMessageService {
         long startTime = System.currentTimeMillis();
 
         try {
+            long absoluteStartTime = System.currentTimeMillis();
             ChatRoom chatRoom = chatRoomRepository.findById(req.roomId())
                     .orElseThrow(() -> new BusinessException(ChatErrorCode.CHAT_ROOM_NOT_FOUND));
             User sender = userRepository.findById(req.senderId())
@@ -389,13 +388,13 @@ public class ChatMessageService {
             ChatRoomSummaryResponse summary = buildChatRoomSummaryResponse(chatRoom.getId(), sender.getId());
 
             // 4. [핵심 변경] 직접 전송하지 않고 이벤트만 던집니다.
-            eventPublisher.publishEvent(new MessageSentEvent(messageResponse, recipientIds, summary));
+            eventPublisher.publishEvent(new MessageSentEvent(messageResponse, recipientIds, summary,absoluteStartTime));
 
-            chatMetrics.onMessageSent("media", true);
-            chatMetrics.recordDelivery(startTime);
+            //chatMetrics.onMessageSent("media", true);
+           // chatMetrics.recordDelivery(startTime);
         } catch (Exception e) {
-            chatMetrics.onMessageSent("media", false);
-            chatMetrics.recordDelivery(startTime);
+           // chatMetrics.onMessageSent("media", false);
+           // chatMetrics.recordDelivery(startTime);
 
             throw e;
         }

--- a/src/main/java/core/domain/notification/service/NotificationEventListener.java
+++ b/src/main/java/core/domain/notification/service/NotificationEventListener.java
@@ -1,5 +1,9 @@
 package core.domain.notification.service;
+import core.domain.userdevicetoken.repository.UserDeviceTokenRepository;
 import core.global.enums.NotificationType;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.domain.Slice;
 import org.springframework.util.StopWatch;
 import core.domain.notification.dto.NewUserJoinedEvent;
 import core.domain.notification.dto.NotificationEvent;
@@ -12,6 +16,8 @@ import org.springframework.scheduling.annotation.Async;
 import org.springframework.stereotype.Component;
 import org.springframework.transaction.annotation.Transactional;
 
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
 import java.util.List;
 
 @Slf4j
@@ -23,7 +29,7 @@ public class NotificationEventListener {
     private final UserNotificationService notificationService;
     private final PushNotificationService pushNotificationService;
     private final NotificationMessageGenerator notificationMessageGenerator;
-
+    private final UserDeviceTokenRepository userDeviceTokenRepository;
 
     @Async("dispatchExecutor")
     @EventListener
@@ -53,57 +59,71 @@ public class NotificationEventListener {
     }
 
     /**
-     * ✅ [비효율적인 버전]
-     * 신규 유저 가입 이벤트를 받아 전체 사용자에게 알림을 발송합니다.
-     * 추후 최적화 필요
+     * ✅ [최적화된 버전]
+     * DB 저장 X, 배치 발송 O, 메모리 효율 O
      */
     @Async("dispatchExecutor")
     @EventListener
-    @Transactional
-    public void handleNewUserBroadcastInefficiently(NewUserJoinedEvent event) {
-        log.info("[비효율적 방식] 신규 유저 가입 이벤트 수신. 전체 알림 발송을 시작합니다. 신규 유저 ID: {}", event.newUserId());
+    public void handleNewUserBroadcastOptimized(NewUserJoinedEvent event) {
 
-        User newUserActor = userRepository.findById(event.newUserId()).orElse(null);
-        if (newUserActor == null) {
-            log.warn("신규 유저 정보를 찾을 수 없어 전체 알림을 중단합니다. ID: {}", event.newUserId());
+        User newUser = userRepository.findById(event.newUserId()).orElse(null);
+        if (newUser == null) {
+            log.warn("신규 유저 정보를 찾을 수 없어 브로드캐스트를 중단합니다. ID: {}", event.newUserId());
             return;
         }
 
-        NotificationEvent tempEvent = new NotificationEvent(
-                null, newUserActor.getId(),
-                NotificationType.newuser,
-                newUserActor.getId(), null, null);
-        String message = notificationMessageGenerator.generateMessage(newUserActor, tempEvent);
-
-        // 1. 전체 유저 조회 (메모리 경고 지점)
-        List<User> allUsers = userRepository.findAll();
-        log.warn("[성능 경고] {}명의 모든 사용자를 메모리에 로드했습니다. 사용자 수가 많을 경우 OutOfMemoryError가 발생할 수 있습니다.", allUsers.size());
-
-        // 2. 시간 측정 시작 (스프링 StopWatch 사용)
+        log.info("📢 신규 유저({}) 브로드캐스트 시작", newUser.getId());
         StopWatch stopWatch = new StopWatch();
         stopWatch.start();
 
-        // 3. 루프 돌며 발송
-        for (User recipient : allUsers) {
-            if (recipient.getId().equals(newUserActor.getId())) {
-                continue;
+        List<String> targetCountries = List.of("South Korea", "KR", "Korea");
+        Instant activeSince = Instant.now().minus(7, ChronoUnit.DAYS);
+        NotificationType notiType = NotificationType.newuser;
+
+        String title = "New friend arrived! 👋";
+        String body = determineMessageInEnglish(newUser);
+
+        int batchSize = 500;
+        Pageable pageable = PageRequest.of(0, batchSize);
+        int totalSent = 0;
+
+        while (true) {
+            Slice<String> tokenSlice = userDeviceTokenRepository.findTokensForBroadcast(
+                    targetCountries, activeSince, notiType, pageable
+            );
+
+            List<String> tokens = tokenSlice.getContent();
+
+            if (!tokens.isEmpty()) {
+                // DB 저장 없이 바로 푸시 발송
+                pushNotificationService.sendBatchPush(tokens, title, body, newUser.getId());
+                totalSent += tokens.size();
             }
 
-            try {
-                notificationService.createAndSaveNotification(recipient, newUserActor, tempEvent, message);
-                pushNotificationService.sendPushNotification(recipient, tempEvent, message);
-            } catch (Exception e) {
-                log.error("사용자 ID {}에게 신규 유저 알림 발송 중 오류 발생", recipient.getId(), e);
-            }
+            if (!tokenSlice.hasNext()) break;
+            pageable = tokenSlice.nextPageable();
         }
 
-        // 4. 시간 측정 종료 및 로그 출력
         stopWatch.stop();
-        log.info("========== [알림 발송 성능 측정] ==========");
+        log.info("========== [브로드캐스트 결과] ==========");
         log.info("총 소요 시간: {} 초", stopWatch.getTotalTimeSeconds());
-        log.info("처리한 유저 수: {} 명", allUsers.size());
+        log.info("발송 건수: {} 건", totalSent);
         log.info("========================================");
+    }
 
-        log.info("[비효율적 방식] 전체 알림 발송 루프 완료. 총 {}명 처리.", allUsers.size());
+    private String determineMessageInEnglish(User newUser) {
+        String country = newUser.getCountry();
+        if (country == null || country.isBlank()) {
+            return "A new friend has joined! Say hello.";
+        }
+
+        country = country.trim();
+        boolean isKorean = country.equalsIgnoreCase("South Korea") || country.equalsIgnoreCase("KR");
+
+        if (isKorean) {
+            return "A new Korean friend has joined! Say hello.";
+        } else {
+            return String.format("A new friend from %s has joined!", country);
+        }
     }
 }

--- a/src/main/java/core/domain/notification/service/NotificationMessageGenerator.java
+++ b/src/main/java/core/domain/notification/service/NotificationMessageGenerator.java
@@ -33,7 +33,6 @@ public class NotificationMessageGenerator {
             case comment -> actorName + " replied to your comment.";
             case follow -> actorName + " accepted your follow request.";
             case receive -> actorName + " started following you.";
-            case newuser -> actorName + " just joined! Say hello!";
             case followuserpost -> actorName + " posted something new.";
             default -> "새로운 알림이 도착했습니다.";
         };

--- a/src/main/java/core/domain/notification/service/PushNotificationService.java
+++ b/src/main/java/core/domain/notification/service/PushNotificationService.java
@@ -121,11 +121,6 @@ public class PushNotificationService {
                             .putData("myId", String.valueOf(recipient.getId()))
                             .putData("roomName", roomName);
                     break;
-                case newuser:
-                    messageBuilder
-                            .putData("type", "newuser")
-                            .putData("userId", String.valueOf(event.referenceId()));
-                    break;
                 case followuserpost:
                     messageBuilder
                             .putData("type", "followuserpost")
@@ -174,20 +169,19 @@ public class PushNotificationService {
     }
     public void sendBatchPush(List<String> tokens, String title, String body, Long actorId) {
         if (tokens.isEmpty()) return;
-
         MulticastMessage message = MulticastMessage.builder()
                 .addAllTokens(tokens)
                 .setNotification(Notification.builder()
                         .setTitle(title)
                         .setBody(body)
                         .build())
+                .putData("notificationType", NotificationType.newuser.name())
                 .putData("type", "newuser")
                 .putData("userId", String.valueOf(actorId))
                 .build();
 
         try {
             BatchResponse response = firebaseMessaging.sendEachForMulticast(message);
-
             if (response.getFailureCount() > 0) {
                 log.warn("배치 발송 완료: 성공 {}, 실패 {}", response.getSuccessCount(), response.getFailureCount());
             }

--- a/src/main/java/core/domain/notification/service/PushNotificationService.java
+++ b/src/main/java/core/domain/notification/service/PushNotificationService.java
@@ -1,9 +1,6 @@
 package core.domain.notification.service;
 
-import com.google.firebase.messaging.FirebaseMessaging;
-import com.google.firebase.messaging.FirebaseMessagingException;
-import com.google.firebase.messaging.Message;
-import com.google.firebase.messaging.MessagingErrorCode;
+import com.google.firebase.messaging.*;
 import core.domain.chat.entity.ChatParticipant;
 import core.domain.chat.repository.ChatParticipantRepository;
 import core.domain.notification.dto.NotificationEvent;
@@ -33,6 +30,7 @@ public class PushNotificationService {
     private final UserNotificationSettingRepository userNotificationSettingRepository;
     private final ChatParticipantRepository chatParticipantRepository;
     private final NotificationMetrics notificationMetrics;
+
 
     /**
      * 사용자에게 푸시 알림을 발송합니다. (람다 제거 버전)
@@ -172,6 +170,29 @@ public class PushNotificationService {
 
                 throw e;
             }
+        }
+    }
+    public void sendBatchPush(List<String> tokens, String title, String body, Long actorId) {
+        if (tokens.isEmpty()) return;
+
+        MulticastMessage message = MulticastMessage.builder()
+                .addAllTokens(tokens)
+                .setNotification(Notification.builder()
+                        .setTitle(title)
+                        .setBody(body)
+                        .build())
+                .putData("type", "newuser")
+                .putData("userId", String.valueOf(actorId))
+                .build();
+
+        try {
+            BatchResponse response = firebaseMessaging.sendEachForMulticast(message);
+
+            if (response.getFailureCount() > 0) {
+                log.warn("배치 발송 완료: 성공 {}, 실패 {}", response.getSuccessCount(), response.getFailureCount());
+            }
+        } catch (FirebaseMessagingException e) {
+            log.error("FCM 배치 발송 중 오류 발생 토큰이 없는 유저", e);
         }
     }
 }

--- a/src/main/java/core/domain/user/controller/MyPageController.java
+++ b/src/main/java/core/domain/user/controller/MyPageController.java
@@ -141,23 +141,6 @@ public class MyPageController {
     }
 
 
-    /**
-     * todo : /profile/edit,profile/skip-setup가 어디서 쓰이는 불명확함 확인 후 스웨거 및 로직 수정해야함
-     * **/
-    @PatchMapping(value = "/profile/skip-setup", consumes = "application/json", produces = "application/json")
-    @Operation(
-            summary = "프로필 셋업 마무리",
-            description = "로그인 후 사용자가 하는 첫 프로필 셋업.사용자가 데이터를 다 넣는다면 USER로 ROLE을 가지고" +
-                    "한개라도 스킵을한다면 ROLE이 VISITOR가 됩니다."
-    )
-    @UserErrorDocs({UserErrorCode.USER_NOT_FOUND})
-    @ImageErrorCodeDocs({ImageErrorCode.IMAGE_UPLOAD_FAILED, ImageErrorCode.IMAGE_FILE_UPLOAD_TYPE_ERROR,  })
-    public ResponseEntity<ApiResponse<LoginResponseDto>> skipSetUpProfile(
-            @Valid @RequestBody UserUpdateDto dto
-    ) {
-        return ResponseEntity.ok(ApiResponse.success(userService.finalizeSkipSetupAndReissueToken(dto)));
-    }
-
     @PatchMapping(value = "/profile/edit", consumes = "application/json", produces = "application/json")
     @Operation(
             summary = "마이페이지 프로필 수정",

--- a/src/main/java/core/domain/user/repository/UserRepository.java
+++ b/src/main/java/core/domain/user/repository/UserRepository.java
@@ -189,16 +189,4 @@ public interface UserRepository extends JpaRepository<User, Long>, UserRepositor
     Object[] visitorShareOverall();
 
 
-    /*
-            * 1. User 엔티티 전체를 가져오지 않고 'fcmToken'만 가져옵니다. (메모리 절약)
-            * 2. DB 인덱스를 활용해 '국가'와 '최근 접속일'로 필터링합니다. (속도 향상)
-            */
-    @Query("SELECT u.fcmToken FROM User u " +
-            "WHERE u.country = :targetCountry " +
-            "AND u.lastSeenAt >= :activeSince")
-    Slice<String> findActiveUserTokens(
-            @Param("targetCountry") String targetCountry,
-            @Param("activeSince") Instant activeSince,
-            Pageable pageable
-    );
 }

--- a/src/main/java/core/domain/user/repository/UserRepository.java
+++ b/src/main/java/core/domain/user/repository/UserRepository.java
@@ -2,6 +2,8 @@ package core.domain.user.repository;
 
 
 import core.domain.user.entity.User;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.domain.Slice;
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.data.jpa.repository.Query;
 import org.springframework.data.repository.query.Param;
@@ -185,4 +187,18 @@ public interface UserRepository extends JpaRepository<User, Long>, UserRepositor
     FROM users
     """, nativeQuery = true)
     Object[] visitorShareOverall();
+
+
+    /*
+            * 1. User 엔티티 전체를 가져오지 않고 'fcmToken'만 가져옵니다. (메모리 절약)
+            * 2. DB 인덱스를 활용해 '국가'와 '최근 접속일'로 필터링합니다. (속도 향상)
+            */
+    @Query("SELECT u.fcmToken FROM User u " +
+            "WHERE u.country = :targetCountry " +
+            "AND u.lastSeenAt >= :activeSince")
+    Slice<String> findActiveUserTokens(
+            @Param("targetCountry") String targetCountry,
+            @Param("activeSince") Instant activeSince,
+            Pageable pageable
+    );
 }

--- a/src/main/java/core/domain/user/service/UserService.java
+++ b/src/main/java/core/domain/user/service/UserService.java
@@ -228,10 +228,28 @@ public class UserService {
         if (dto.imageKey() != null) {
             imageService.saveUserProfileImage(user.getId(), dto.imageKey());
         }
-
-        //publisher.publishEvent(new NewUserJoinedEvent(user.getId()));
+        if (isAllProfileFieldsFilled(dto)) {
+            publisher.publishEvent(new NewUserJoinedEvent(user.getId()));
+            log.info("모든 프로필 정보 입력 완료. 신규 유저 알림 이벤트 발행: UserID={}", user.getId());
+        } else {
+            log.info("일부 프로필 정보 누락으로 알림 미발행: {}", dto);
+        }
     }
-
+    /**
+     * DTO의 필수 필드가 모두 채워졌는지 검사하는 메서드
+     */
+    private boolean isAllProfileFieldsFilled(UserSetupRequest dto) {
+        return notBlank(dto.firstname()) &&
+                notBlank(dto.lastname()) &&
+                notBlank(dto.gender()) &&
+                notBlank(dto.birthday()) &&
+                notBlank(dto.country()) &&
+                notBlank(dto.introduction()) &&
+                notBlank(dto.purpose()) &&
+                notBlank(dto.imageKey()) && // 이미지도 필수
+                dto.language() != null && !dto.language().isEmpty() && // 언어도 1개 이상
+                dto.hobby() != null && !dto.hobby().isEmpty(); // 취미도 1개 이상
+    }
     private boolean notBlank(String s) {
         return s != null && !s.isBlank();
     }

--- a/src/main/java/core/domain/userdevicetoken/repository/UserDeviceTokenRepository.java
+++ b/src/main/java/core/domain/userdevicetoken/repository/UserDeviceTokenRepository.java
@@ -2,8 +2,14 @@ package core.domain.userdevicetoken.repository;
 
 import core.domain.user.entity.User;
 import core.domain.userdevicetoken.entity.UserDeviceToken;
+import core.global.enums.NotificationType;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.domain.Slice;
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
 
+import java.time.Instant;
 import java.util.List;
 import java.util.Optional;
 
@@ -14,4 +20,27 @@ public interface UserDeviceTokenRepository extends JpaRepository<UserDeviceToken
     List<UserDeviceToken> findAllByUserId(Long id);
     void deleteAllByUserId(Long userId);
     List<UserDeviceToken> findAllByUser(User user);
+
+    /**
+     * 🚀 [최적화 쿼리]
+     * 1. Target: 'UserDeviceToken' 테이블 기준
+     * 2. Join: User(국가, 활동시간 확인용) + UserNotificationSetting(알림 수신 동의 확인용)
+     * 3. Filter: 한국인(KR) + 최근 7일 활동 + 알림 설정 ON
+     * 4. Select: 엔티티가 아닌 'deviceToken(String)'만 조회 (메모리 절약)
+     */
+    @Query("SELECT udt.deviceToken " +
+            "FROM UserDeviceToken udt " +
+            "JOIN udt.user u " +
+            "JOIN u.notificationSettings uns " +
+            "WHERE u.country IN :targetCountries " +  // 1. 타겟 국가 (한국)
+            "AND u.lastSeenAt >= :activeSince " +     // 2. 최근 활동 유저
+            "AND uns.notificationType = :type " +     // 3. 알림 타입 (NEW_USER)
+            "AND uns.enabled = true " +               // 4. 알림 켜둔 사람만
+            "AND udt.deviceToken IS NOT NULL")
+    Slice<String> findTokensForBroadcast(
+            @Param("targetCountries") List<String> targetCountries,
+            @Param("activeSince") Instant activeSince,
+            @Param("type") NotificationType type,
+            Pageable pageable
+    );
 }

